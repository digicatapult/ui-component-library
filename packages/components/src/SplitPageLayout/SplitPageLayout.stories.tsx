import React from 'react'
import {Meta, Story} from '@storybook/react'
import {Box, Button, Heading, Text} from '..'
import {NavList} from '../NavList'
import {Placeholder} from '../Placeholder'
import {SplitPageLayout} from '../SplitPageLayout'

const meta: Meta = {
  title: 'Components/SplitPageLayout',
  parameters: {
    layout: 'fullscreen',
    controls: {expanded: true},
  },
  argTypes: {
    // Debug controls
    'Render header?': {
      type: 'boolean',
      defaultValue: true,
      table: {category: 'Debug'},
    },
    'Render pane?': {
      type: 'boolean',
      defaultValue: true,
      table: {category: 'Debug'},
    },
    'Render footer?': {
      type: 'boolean',
      defaultValue: true,
      table: {category: 'Debug'},
    },
    'Header placeholder height': {
      type: 'number',
      defaultValue: 64,
      table: {category: 'Debug'},
    },
    'Pane placeholder height': {
      type: 'number',
      defaultValue: 200,
      table: {category: 'Debug'},
    },
    'Content placeholder height': {
      type: 'number',
      defaultValue: 400,
      table: {category: 'Debug'},
    },
    'Footer placeholder height': {
      type: 'number',
      defaultValue: 64,
      table: {category: 'Debug'},
    },

    // Header prop controls
    'Header.divider.regular': {
      type: {
        name: 'enum',
        value: ['none', 'line'],
      },
      defaultValue: 'line',
      control: {
        type: 'radio',
      },
      table: {
        category: 'Header props',
      },
    },
    'Header.divider.narrow': {
      type: {
        name: 'enum',
        value: ['none', 'line', 'filled'],
      },
      defaultValue: 'line',
      control: {
        type: 'radio',
      },
      table: {
        category: 'Header props',
      },
    },
    'Header.divider.wide': {
      type: {
        name: 'enum',
        value: ['none', 'line'],
      },
      defaultValue: 'line',
      control: {
        type: 'radio',
      },
      table: {
        category: 'Header props',
      },
    },
    'Header.padding': {
      type: {
        name: 'enum',
        value: ['none', 'condensed', 'normal'],
      },
      defaultValue: 'normal',
      control: {type: 'radio'},
      table: {category: 'Header props'},
    },
    'Header.hidden.regular': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Header props'},
    },
    'Header.hidden.narrow': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Header props'},
    },
    'Header.hidden.wide': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Header props'},
    },

    // Content prop controls
    'Content.width': {
      type: {
        name: 'enum',
        value: ['full', 'medium', 'large', 'xlarge'],
      },
      defaultValue: 'xlarge',
      control: {type: 'radio'},
      table: {category: 'Content props'},
    },
    'Content.padding': {
      type: {
        name: 'enum',
        value: ['none', 'condensed', 'normal'],
      },
      defaultValue: 'normal',
      control: {type: 'radio'},
      table: {category: 'Content props'},
    },
    'Content.hidden.regular': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Content props'},
    },
    'Content.hidden.narrow': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Content props'},
    },
    'Content.hidden.wide': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Content props'},
    },

    // Pane prop controls
    'Pane.position.regular': {
      type: {
        name: 'enum',
        value: ['start', 'end'],
      },
      defaultValue: 'start',
      control: {type: 'radio'},
      table: {category: 'Pane props'},
    },
    'Pane.position.narrow': {
      type: {
        name: 'enum',
        value: ['start', 'end'],
      },
      defaultValue: 'start',
      control: {type: 'radio'},
      table: {category: 'Pane props'},
    },
    'Pane.position.wide': {
      type: {
        name: 'enum',
        value: ['start', 'end'],
      },
      defaultValue: 'start',
      control: {type: 'radio'},
      table: {category: 'Pane props'},
    },
    'Pane.width': {
      type: {
        name: 'enum',
        value: ['small', 'medium', 'large'],
      },
      defaultValue: 'medium',
      control: {type: 'radio'},
      table: {category: 'Pane props'},
    },
    'Pane.sticky': {
      type: 'boolean',
      defaultValue: true,
      table: {category: 'Pane props'},
    },
    'Pane.resizable': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Pane props'},
    },
    'Pane.padding': {
      type: {
        name: 'enum',
        value: ['none', 'condensed', 'normal'],
      },
      defaultValue: 'normal',
      control: {type: 'radio'},
      table: {category: 'Pane props'},
    },
    'Pane.divider.regular': {
      type: {
        name: 'enum',
        value: ['none', 'line'],
      },
      defaultValue: 'line',
      control: {type: 'radio'},
      table: {category: 'Pane props'},
    },
    'Pane.divider.narrow': {
      type: {
        name: 'enum',
        value: ['none', 'line', 'filled'],
      },
      defaultValue: 'line',
      control: {type: 'radio'},
      table: {category: 'Pane props'},
    },
    'Pane.divider.wide': {
      type: {
        name: 'enum',
        value: ['none', 'line'],
      },
      defaultValue: 'line',
      control: {type: 'radio'},
      table: {category: 'Pane props'},
    },

    // Footer prop controls
    'Footer.divider.regular': {
      type: {
        name: 'enum',
        value: ['none', 'line'],
      },
      defaultValue: 'line',
      control: {
        type: 'radio',
      },
      table: {
        category: 'Footer props',
      },
    },
    'Footer.divider.narrow': {
      type: {
        name: 'enum',
        value: ['none', 'line', 'filled'],
      },
      defaultValue: 'line',
      control: {
        type: 'radio',
      },
      table: {
        category: 'Footer props',
      },
    },
    'Footer.divider.wide': {
      type: {
        name: 'enum',
        value: ['none', 'line'],
      },
      defaultValue: 'line',
      control: {
        type: 'radio',
      },
      table: {
        category: 'Footer props',
      },
    },
    'Footer.padding': {
      type: {
        name: 'enum',
        value: ['none', 'condensed', 'normal'],
      },
      defaultValue: 'normal',
      control: {type: 'radio'},
      table: {category: 'Footer props'},
    },
    'Footer.hidden.regular': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Footer props'},
    },
    'Footer.hidden.narrow': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Footer props'},
    },
    'Footer.hidden.wide': {
      type: 'boolean',
      defaultValue: false,
      table: {category: 'Footer props'},
    },
  },
}

const Template: Story = args => (
  <SplitPageLayout sx={args.sx}>
    {args['Render header?'] ? (
      <SplitPageLayout.Header
        padding={args['Header.padding']}
        divider={{
          narrow: args['Header.divider.narrow'],
          regular: args['Header.divider.regular'],
          wide: args['Header.divider.wide'],
        }}
        hidden={{
          narrow: args['Header.hidden.narrow'],
          regular: args['Header.hidden.regular'],
          wide: args['Header.hidden.wide'],
        }}
      >
        <Placeholder height={args['Header placeholder height']} label="Header" />
      </SplitPageLayout.Header>
    ) : null}
    <SplitPageLayout.Content
      width={args['Content.width']}
      padding={args['Content.padding']}
      hidden={{
        narrow: args['Content.hidden.narrow'],
        regular: args['Content.hidden.regular'],
        wide: args['Content.hidden.wide'],
      }}
    >
      <Placeholder height={args['Content placeholder height']} label="Content" />
    </SplitPageLayout.Content>
    {args['Render pane?'] ? (
      <SplitPageLayout.Pane
        resizable={args['Pane.resizable']}
        position={{
          narrow: args['Pane.position.narrow'],
          regular: args['Pane.position.regular'],
          wide: args['Pane.position.wide'],
        }}
        width={args['Pane.width']}
        sticky={args['Pane.sticky']}
        padding={args['Pane.padding']}
        divider={{
          narrow: args['Pane.divider.narrow'],
          regular: args['Pane.divider.regular'],
          wide: args['Pane.divider.wide'],
        }}
        hidden={{
          narrow: args['Pane.hidden.narrow'],
          regular: args['Pane.hidden.regular'],
          wide: args['Pane.hidden.wide'],
        }}
      >
        <Placeholder height={args['Pane placeholder height']} label="Pane" />
      </SplitPageLayout.Pane>
    ) : null}
    {args['Render footer?'] ? (
      <SplitPageLayout.Footer
        padding={args['Footer.padding']}
        divider={{
          narrow: args['Footer.divider.narrow'],
          regular: args['Footer.divider.regular'],
          wide: args['Footer.divider.wide'],
        }}
        hidden={{
          narrow: args['Footer.hidden.narrow'],
          regular: args['Footer.hidden.regular'],
          wide: args['Footer.hidden.wide'],
        }}
      >
        <Placeholder height={args['Footer placeholder height']} label="Footer" />
      </SplitPageLayout.Footer>
    ) : null}
  </SplitPageLayout>
)

export const Default = Template.bind({})

export const SettingsPage = () => (
  <SplitPageLayout>
    <SplitPageLayout.Pane position="start">
      <NavList aria-label="main">
        <NavList.Item href="#">Profile</NavList.Item>
        <NavList.Item href="#" aria-current="page">
          Account
        </NavList.Item>
        <NavList.Item href="#">Emails</NavList.Item>
        <NavList.Item href="#">Notifications</NavList.Item>
      </NavList>
    </SplitPageLayout.Pane>
    <SplitPageLayout.Content>
      <Heading as="h2" sx={{fontSize: 4, fontWeight: 'normal', color: 'danger.fg', mb: 2}}>
        Danger zone
      </Heading>
      <Box
        sx={{
          border: '1px solid',
          borderColor: 'danger.emphasis',
          borderRadius: 2,
          p: 3,
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'space-between',
          gap: 3,
        }}
      >
        <Box sx={{display: 'grid', gap: 1}}>
          <Text sx={{fontSize: 1, fontWeight: 'bold', color: 'danger.fg'}}>Delete account</Text>
          <Text sx={{fontSize: 1, color: 'fg.muted'}}>
            Are you sure you don&apos;t want to just downgrade your account to a free account? We won&apos;t charge your
            credit card anymore.
          </Text>
        </Box>
        <Button variant="danger">Delete account</Button>
      </Box>
    </SplitPageLayout.Content>
  </SplitPageLayout>
)

export default meta
