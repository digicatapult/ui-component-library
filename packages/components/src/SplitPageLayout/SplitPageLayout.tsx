import React from 'react'
import {SxProp} from '..'
import {
  PageLayout,
  PageLayoutContentProps,
  PageLayoutFooterProps,
  PageLayoutHeaderProps,
  PageLayoutPaneProps,
} from '../PageLayout'

// ----------------------------------------------------------------------------
// SplitPageLayout

export type SplitPageLayoutProps = SxProp

export const Root: React.FC<React.PropsWithChildren<SplitPageLayoutProps>> = props => {
  return <PageLayout containerWidth="full" padding="none" columnGap="none" rowGap="none" {...props} />
}

Root.displayName = 'SplitPageLayout'

// ----------------------------------------------------------------------------
// SplitPageLayout.Header

export type SplitPageLayoutHeaderProps = PageLayoutHeaderProps

export const Header: React.FC<React.PropsWithChildren<SplitPageLayoutHeaderProps>> = ({
  padding = 'normal',
  divider = 'line',
  ...props
}) => {
  return <PageLayout.Header padding={padding} divider={divider} {...props} />
}

Header.displayName = 'SplitPageLayout.Header'

// ----------------------------------------------------------------------------
// SplitPageLayout.Content

export type SplitPageLayoutContentProps = PageLayoutContentProps

export const Content: React.FC<React.PropsWithChildren<SplitPageLayoutContentProps>> = ({
  width = 'large',
  padding = 'normal',
  ...props
}) => {
  return <PageLayout.Content width={width} padding={padding} {...props} />
}

Content.displayName = 'SplitPageLayout.Content'

// ----------------------------------------------------------------------------
// SplitPageLayout.Pane

export type SplitPageLayoutPaneProps = PageLayoutPaneProps

export const Pane: React.FC<React.PropsWithChildren<SplitPageLayoutPaneProps>> = ({
  position = 'start',
  sticky = true,
  padding = 'normal',
  divider = 'line',
  ...props
}) => {
  return <PageLayout.Pane position={position} sticky={sticky} padding={padding} divider={divider} {...props} />
}
Pane.displayName = 'SplitPageLayout.Pane'

// ----------------------------------------------------------------------------
// SplitPageLayout.Footer

export type SplitPageLayoutFooterProps = PageLayoutFooterProps

export const Footer: React.FC<React.PropsWithChildren<SplitPageLayoutFooterProps>> = ({
  padding = 'normal',
  divider = 'line',
  ...props
}) => {
  return <PageLayout.Footer padding={padding} divider={divider} {...props} />
}

Footer.displayName = 'SplitPageLayout.Footer'

// ----------------------------------------------------------------------------
// Export

export const SplitPageLayout = Object.assign(Root, {
  Header,
  Content,
  Pane,
  Footer,
})
