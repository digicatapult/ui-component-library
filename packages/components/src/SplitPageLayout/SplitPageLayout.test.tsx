import React from 'react'
import {render} from '@testing-library/react'
import MatchMediaMock from 'jest-matchmedia-mock'
import 'react-intersection-observer/test-utils'
import {ThemeProvider} from '..'
import {SplitPageLayout} from '../SplitPageLayout/SplitPageLayout'

let matchMedia: MatchMediaMock

describe('SplitPageLayout', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock()
  })

  afterEach(() => {
    matchMedia.clear()
  })

  it('renders default layout', () => {
    const {container} = render(
      <ThemeProvider>
        <SplitPageLayout>
          <SplitPageLayout.Header>Header</SplitPageLayout.Header>
          <SplitPageLayout.Content>Content</SplitPageLayout.Content>
          <SplitPageLayout.Pane>Pane</SplitPageLayout.Pane>
          <SplitPageLayout.Footer>Footer</SplitPageLayout.Footer>
        </SplitPageLayout>
      </ThemeProvider>,
    )
    expect(container).toMatchSnapshot()
  })
})
