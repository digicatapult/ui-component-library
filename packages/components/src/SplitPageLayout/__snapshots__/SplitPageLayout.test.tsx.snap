// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitPageLayout renders default layout 1`] = `
.c0 {
  padding: 0;
}

.c1 {
  max-width: 100%;
  margin-left: auto;
  margin-right: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 100%;
  -ms-flex: 1 1 100%;
  flex: 1 1 100%;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  max-width: 100%;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  min-width: 1px;
}

.c7 {
  width: 100%;
  max-width: 1012px;
  margin-left: auto;
  margin-right: auto;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 16px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  width: 100%;
  margin-left: 0;
  margin-right: 0;
  -webkit-flex-direction: column-reverse;
  -ms-flex-direction: column-reverse;
  flex-direction: column-reverse;
  margin-bottom: 0;
}

.c9 {
  margin-left: 0;
  margin-right: 0;
  display: none;
  margin-top: 0;
}

.c10 {
  height: 100%;
  position: relative;
  display: none;
  margin-left: 0;
}

.c11 {
  --pane-min-width: 256px;
  --pane-max-width: calc(100vw - 511px);
  width: 100%;
  padding: 16px;
}

.c2 {
  width: 100%;
  margin-bottom: 0;
}

.c3 {
  padding: 16px;
}

.c4 {
  margin-left: 0;
  margin-right: 0;
  display: block;
  height: 1px;
  background-color: #d0d7de;
  margin-top: 0;
}

.c12 {
  -webkit-order: 4;
  -ms-flex-order: 4;
  order: 4;
  width: 100%;
  margin-top: 0;
}

.c13 {
  margin-left: 0;
  margin-right: 0;
  display: block;
  height: 1px;
  background-color: #d0d7de;
  margin-bottom: 0;
}

@media screen and (min-width:1012px) {
  .c7 {
    padding: 24px;
  }
}

@media screen and (min-width:768px) {
  .c8 {
    width: auto;
    margin-top: 0 !important;
    margin-bottom: 0 !important;
    position: -webkit-sticky;
    position: sticky;
    top: 0px;
    max-height: var(--sticky-pane-height);
    -webkit-flex-direction: row-reverse;
    -ms-flex-direction: row-reverse;
    flex-direction: row-reverse;
    margin-right: 0;
  }
}

@media screen and (min-width:768px) {
  .c9 {
    margin-left: 0 !important;
    margin-right: 0 !important;
  }
}

@media screen and (min-width:768px) {
  .c11 {
    overflow: auto;
  }
}

@media screen and (min-width:1012px) {
  .c11 {
    width: 296px;
    padding: 24px;
  }
}

@media screen and (min-width:1280px) {
  .c11 {
    --pane-max-width: calc(100vw - 959px);
  }
}

@media screen and (min-width:1012px) {
  .c3 {
    padding: 24px;
  }
}

@media screen and (min-width:768px) {
  .c4 {
    margin-left: 0 !important;
    margin-right: 0 !important;
  }
}

@media screen and (min-width:768px) {
  .c13 {
    margin-left: 0 !important;
    margin-right: 0 !important;
  }
}

<div>
  <div
    class="c0"
    style="--sticky-pane-height: 100vh;"
  >
    <div
      class="c1"
    >
      <header
        class="c2"
      >
        <div
          class="c3"
        >
          Header
        </div>
        <div
          class="c4"
        />
      </header>
      <div
        class="c5"
      >
        <main
          class="c6"
        >
          <div
            class=""
          />
          <div
            class="c7"
          >
            Content
          </div>
          <div
            class=""
          />
        </main>
        <div
          class="c8"
        >
          <div
            class="c9"
          />
          <div
            class="c10"
          />
          <div
            class="c11"
            style="--pane-width: 296px;"
          >
            Pane
          </div>
        </div>
      </div>
      <footer
        class="c12"
      >
        <div
          class="c13"
        />
        <div
          class="c3"
        >
          Footer
        </div>
      </footer>
    </div>
  </div>
</div>
`;
